### ISSUE TITLE
Errors in before/finally blocks cause runner to crash

### ISSUE THREAD

**Issue by jonathanperret** [antonmi/espec#81]
Given the following spec:

```elixir
defmodule BeforeIssue do
  use ESpec

  before do
    raise "some problem"
  end

  it do
    expect(1).to eq(1)
  end
end
```

I would expect Espec to report that the `before` block failed to run (and ideally, report that one example could not be run).

Instead, the example runner crashes:

```
=ERROR REPORT==== 17-Feb-2016::17:44:13 ===
** Generic server 'Elixir.ESpec.Runner' terminating
** Last message in was run
** When Server state == #{opts => [{finally,#Fun<elixir_compiler_1.2.64185228>},
                           {finish_loading_time,{1455,727453,597612}},
                           {file_opts,[{<<"/Users/jonathanperret/src/espec/before_issue.exs">>,
                                        []}]},
                           {start_loading_time,{1455,727453,510711}},
                           {before,#Fun<elixir_compiler_1.1.64185228>},
                           {included_applications,[]}],
                          specs => ['Elixir.BeforeIssue']}
** Reason for termination ==
** {#{'__exception__' => true,
      '__struct__' => 'Elixir.RuntimeError',
      message => <<"some problem">>},
    [{'Elixir.BeforeIssue',before_kdqpgulzhjvxtcwsmyebnaifro,1,
                           [{file,"before_issue.exs"},{line,5}]},
     {'Elixir.ESpec.ExampleRunner','-run_befores_and_lets/2-fun-0-',2,
                                   [{file,"lib/espec/example_runner.ex"},
                                    {line,102}]},
     {'Elixir.Enum','-reduce/3-lists^foldl/2-0-',3,
                    [{file,"lib/enum.ex"},{line,1473}]},
     {'Elixir.ESpec.ExampleRunner',run_example,2,
                                   [{file,"lib/espec/example_runner.ex"},
                                    {line,31}]},
     {'Elixir.Enum','-map/2-lists^map/1-0-',2,
                    [{file,"lib/enum.ex"},{line,1088}]},
     {'Elixir.ESpec.Runner',run_examples,1,
                            [{file,"lib/espec/runner.ex"},{line,69}]},
     {'Elixir.ESpec.Runner',do_run,2,[{file,"lib/espec/runner.ex"},{line,42}]},
     {'Elixir.ESpec.Runner',handle_call,3,
                            [{file,"lib/espec/runner.ex"},{line,28}]}]}
** (EXIT from #PID<0.47.0>) an exception was raised:
    ** (RuntimeError) some problem
        before_issue.exs:5: BeforeIssue.before_kdqpgulzhjvxtcwsmyebnaifro/1
        (espec) lib/espec/example_runner.ex:102: anonymous fn/2 in ESpec.ExampleRunner.run_befores_and_lets/2
        (elixir) lib/enum.ex:1473: Enum."-reduce/3-lists^foldl/2-0-"/3
        (espec) lib/espec/example_runner.ex:31: ESpec.ExampleRunner.run_example/2
        (elixir) lib/enum.ex:1088: Enum."-map/2-lists^map/1-0-"/2
        (espec) lib/espec/runner.ex:69: ESpec.Runner.run_examples/1
        (espec) lib/espec/runner.ex:42: ESpec.Runner.do_run/2
        (espec) lib/espec/runner.ex:28: ESpec.Runner.handle_call/3
```

The same problem exists with `finally` blocks.

As well as raised errors, thrown terms should probably be handled as well, as per #80.

**Comment by antonmi** [comment]
Thanks, Jonathan!

**Comment by antonmi** [comment]
The fix will be available in the next ESpec version

**Comment by antonmi** [comment]
The fix is included into 0.8.12

**Comment by jonathanperret** [comment]
Thanks!
